src/Core/Content/Media/Infrastructure/Path/MediaUrlGenerator.php

```diff
@@ -33,9 +33,45 @@ public function generate(array $paths): array
                 $url .= '?ts=' . $value->updatedAt->getTimestamp();
             }
 
-            $urls[$key] = $url;
+            $urls[$key] = $this->encodeUrl($url);
         }
 
         return $urls;
     }
+
+    private function encodeUrl(string $mediaUrl): string
+    {
+        $urlInfo = parse_url($mediaUrl);
+        if (!\is_array($urlInfo)) {
+            return $mediaUrl;
+        }
+
+        $segments = explode('/', $urlInfo['path'] ?? '');
+
+        foreach ($segments as $index => $segment) {
+            $segments[$index] = rawurlencode($segment);
+        }
+
+        $path = implode('/', $segments);
+        if (isset($urlInfo['query'])) {
+            $path .= "?{$urlInfo['query']}";
+        }
+
+        $encodedPath = '';
+
+        if (isset($urlInfo['scheme'])) {
+            $encodedPath = "{$urlInfo['scheme']}://";
+        }
+
+        if (isset($urlInfo['host'])) {
+            $encodedPath .= "{$urlInfo['host']}";
+        }
+
+        if (isset($urlInfo['port'])) {
```

Comment:
Maybe I am missing something, but is it possible to have a port without a host? And will the result still be valid?